source/unified-test-format/unified-test-format.rst

```diff
@@ -0,0 +1,1896 @@
+setup, but the entity map may also be modified during test execution via the
+`operation.saveResultAsEntity <operation_saveResultAsEntity_>`_ directive.
+
+Test runners MAY choose to implement the entity map in a fashion most suited to
+their language, but implementations MUST enforce both uniqueness of entity names
+and referential integrity when fetching an entity. Test runners MUST raise an
+error if an attempt is made to store an entity with a name that already exists
+in the map and MUST raise an error if an entity is not found for a name or is
+found but has an unexpected type.
+
+Consider the following examples::
+
+    # Error due to a duplicate name (client0 was already defined)
+    createEntities:
+      - client: { id: client0 }
+      - client: { id: client0 }
+
+    # Error due to a missing entity (client1 is undefined)
+    createEntities:
+      - client: { id: client0 }
+      - session: { id: session0, client: client1 }
+
+    # Error due to an unexpected entity type (session instead of client)
+    createEntities:
+      - client: { id: client0 }
+      - session: { id: session0, client: client0 }
+      - session: { id: session1, client: session0 }
+
+
+Test Format
+-----------
+
+Each specification test file can define one or more tests, which inherit some
+top-level configuration (e.g. namespace, initial data). YAML and JSON test files
+are parsed as a document by the test runner. This section defines the top-level
+keys for that document and links to various sub-sections for definitions of
+nested structures (e.g. individual `test`_, `operation`_).
+
+Although test runners are free to process YAML or JSON files, YAML is the
+canonical format for writing tests. YAML files may be converted to JSON using a
+tool such as `js-yaml <https://github.com/nodeca/js-yaml>`__ .
+
+
+Top-level Fields
+~~~~~~~~~~~~~~~~
+
+The top-level fields of a test file are as follows:
+
+.. _schemaVersion:
+
+- ``schemaVersion``: Required string. Version of this specification to which the
+  test file complies. Test runners will use this to determine compatibility
+  (i.e. whether and how the test file will be interpreted). The format of this
+  string is defined in `Version String`_.
+
+.. _runOn:
+
+- ``runOn``: Optional array of documents. List of server version and/or topology
+  requirements for which the tests in this file can be run. These requirements
+  may be overridden on a per-test basis by `test.runOn <test_runOn_>`_. Test
+  runners MUST skip a test if its requirements are not met.
+
+  If set, the array should contain at least one document. The structure of each
+  document is defined in `runOnRequirement`_.
+
+  **TODO**: Consider whether these requirements should allow an entire test file
+  to be skipped up front, instead of allowing per-test requirements to override.
+
+.. _allowMultipleMongoses:
+
+- ``allowMultipleMongoses``: Optional boolean. If false, all MongoClients
```

Comment:
I was under the impression that this needed to apply to the internal MongoClient, but I may have gotten that wrong when porting this over from transactions.

As-is, this being a top-level option means a single test file cannot meaningfully utilize both `failPoint` and `targetedFailPoint` in different tests. Moving it back to either the `test` block or client entity definition would remove that restriction, and I do agree that the client definition makes more sense (even if that means some files might create two clients and only use one at a time).

Reply:
Will move this to client entity definition, since it's only relevant for `targetedFailtPoint` usage.

The internal MongoClient should always connect to all available mongos (needed to kill sessions).

We also need to change `failPoint` to use a client entity, which can be configured to restrict to a single mongos.

Reply:
`allowMultipleMongoses` is now client-level `useMultipleMongoses` and I've updated `failPoint` to use a client entity. See changelog.